File: MaiChat/desktop/src/model/MessageQuote.cpp
#include "model/MessageQuote.h"

#include <QFileInfo>

namespace MessageQuote {

namespace {

// 入站附件消息的 text 有时是内部占位串（「[图片消息] xxx」这种），那不是用户写的配文。
// 拿它当摘要会把内部约定漏进引用块里。判据与 MessageNotification 保持一致。
bool looksLikePlaceholderText(const QString& text) {
    return text.startsWith(QStringLiteral("[图片消息] "))
        || text.startsWith(QStringLiteral("[文件消息] "))
        || text.startsWith(QStringLiteral("[视频消息] "))
        || text.startsWith(QStringLiteral("[语音消息] "));
}

// 用户真正写的配文；没有就返回空串。
QString authoredCaption(const RemoteIMMessage& message) {
    const QString text = message.text.trimmed();
    if (text.isEmpty() || looksLikePlaceholderText(text)) return QString();
    return text;
}

// simplified() 正好是「去首尾空白 + 连续空白折成单个空格」，换行也一并折掉：
// 引用块是单行展示，原文里的换行带进来会把布局撑开。
QString clamped(const QString& text) {
    const QString clean = text.simplified();
    if (clean.size() <= kDigestLimit) return clean;
    return clean.left(kDigestLimit) + QStringLiteral("…");
}

}  // namespace

QString kindOf(const RemoteIMMessage& message) {
    if (message.hasImage) return QStringLiteral("image");
    if (message.hasVideo) return QStringLiteral("video");
    if (message.hasVoice) return QStringLiteral("voice");
    if (message.hasFile) return QStringLiteral("file");
    return QStringLiteral("text");
}

QString digestOf(const RemoteIMMessage& message) {
    const QString caption = authoredCaption(message);
    if (message.hasImage) {
        return clamped(caption.isEmpty() ? QStringLiteral("[图片]") : caption);
    }
    if (message.hasVideo) {
        return clamped(caption.isEmpty() ? QStringLiteral("[视频]") : caption);
    }
    if (message.hasVoice) {
        return clamped(caption.isEmpty() ? QStringLiteral("[语音]") : caption);
    }
    if (message.hasFile) {
        if (!caption.isEmpty()) return clamped(caption);
        // 文件名只取 basename。file.name 一般已经是文件名，但它也可能被写成路径，
        // 而把本地路径放进要发给对端的消息里是不能接受的。
        const QString name = QFileInfo(message.file.fileName.isEmpty() ? message.file.localPath
                                                                      : message.file.fileName)
                                 .fileName()
                                 .trimmed();
        return clamped(name.isEmpty() ? QStringLiteral("[文件]")
                                      : QStringLiteral("[文件] ") + name);
    }
    return clamped(caption);
}

QString sdkMessageIdOf(const QString& localMessageId) {
    const int hash = localMessageId.lastIndexOf(QLatin1Char('#'));
    // 只认「末尾是 #<十进制数字>」这一种形状。按第一个 # 截断是不行的：
    // 协议层不该偷偷假设 SDK id 自身永远不含 #。
    if (hash <= 0 || hash == localMessageId.size() - 1) return QString();
    const QStringView suffix = QStringView(localMessageId).mid(hash + 1);
    for (const QChar c : suffix) {
        if (!c.isDigit()) return QString();
    }
    return localMessageId.left(hash);
}

RemoteIMQuote quoteFor(const RemoteIMMessage& message) {
    RemoteIMQuote quote;
    // 拿不到 SDK 派生 ID 就留空。随机 UUID 跨端无法解析，
    // 发出去只会让对端拿到一个永远命中不了的 ID。
    quote.msgId = sdkMessageIdOf(message.id);
    quote.senderId = message.fromUserId;
    quote.digest = digestOf(message);
    quote.kind = kindOf(message);
    return quote;
}

}  // namespace MessageQuote

File: MaiChat/android/MaiChat/app/src/main/java/com/kongshang/maichat/MessageQuote.java
package com.kongshang.maichat;

import java.io.File;

public final class MessageQuote {
    static final int DIGEST_LIMIT = 120;

    private MessageQuote() {
    }

    public static RemoteIMQuote from(RemoteIMMessage message) {
        if (message == null) throw new IllegalArgumentException("message is required");
        return new RemoteIMQuote(
            message.remoteId(),
            message.fromUserId(),
            digest(message),
            kind(message)
        );
    }

    public static String kind(RemoteIMMessage message) {
        if (message.imageAttachment() != null) return "image";
        if (message.videoAttachment() != null) return "video";
        if (message.voiceAttachment() != null) return "voice";
        if (message.fileAttachment() != null) return "file";
        return "text";
    }

    public static String digest(RemoteIMMessage message) {
        String caption = authoredCaption(message.text());
        if (message.imageAttachment() != null) return clamp(caption.isEmpty() ? "[图片]" : caption);
        if (message.videoAttachment() != null) return clamp(caption.isEmpty() ? "[视频]" : caption);
        if (message.voiceAttachment() != null) return clamp(caption.isEmpty() ? "[语音]" : caption);
        if (message.fileAttachment() != null) {
            if (!caption.isEmpty()) return clamp(caption);
            String rawName = message.fileAttachment().fileName();
            if (rawName == null || rawName.trim().isEmpty()) rawName = message.fileAttachment().localPath();
            String name = basename(rawName);
            return clamp(name.isEmpty() ? "[文件]" : "[文件] " + name);
        }
        return clamp(caption);
    }

    private static String authoredCaption(String text) {
        String clean = clean(text);
        if (clean.startsWith("[图片消息] ")
            || clean.startsWith("[文件消息] ")
            || clean.startsWith("[视频消息] ")
            || clean.startsWith("[语音消息] ")) return "";
        return clean;
    }

    private static String basename(String value) {
        String normalized = clean(value).replace('\\', '/');
        int slash = normalized.lastIndexOf('/');
        String name = slash >= 0 ? normalized.substring(slash + 1) : normalized;
        return new File(name).getName().trim();
    }

    private static String clamp(String value) {
        String clean = clean(value).replaceAll("\\s+", " ");
        if (clean.length() <= DIGEST_LIMIT) return clean;
        return clean.substring(0, DIGEST_LIMIT) + "…";
    }

    private static String clean(String value) {
        return value == null ? "" : value.trim();
    }
}
